### PR TITLE
Update CMake paths for dependent package use

### PR DESCRIPTION
**Pull request by mlxd**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:** Currently the new package design assumes that all use will be internal to the nested projects. This however prevents packages depending on Lightning, for example through the CMake FetchContent utility, as source directories are assumed to be top-level of the outer package only. This PR updates all Lightning CMake references from `CMAKE_SOURCE_DIR` to `pennylane_lightning_SOURCE_DIR`, as discussed by https://cmake.org/cmake/help/v3.27/command/project.html#command:project

**Description of the Change:** As above.

**Benefits:** Allows the repo to be used as a sub-package with CMake.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -21,6 +21,9 @@
 
 ### Bug fixes
 
+* Update the CMake interal references to enable sub-project compilation with affecting the parent package.
```
Comment (mlxd) [inline .github/CHANGELOG.md:24]:
```suggestion
* Update the CMake internal references to enable sub-project compilation with affecting the parent package.
```

File: .github/CHANGELOG.md [hunk 2]
```diff
@@ -21,6 +21,9 @@
 
 ### Bug fixes
 
+* Update the CMake internal references to enable sub-project compilation with affecting the parent package.
```
Comment (AmintorDusko) [inline .github/CHANGELOG.md:24]: I know why you did it, but I'm not getting what you meant by "with affecting the parent package.".
Could you maybe rephrase that?

**Review by vincentmr — APPROVED** [review]
Unblocks LGPU updates. Thanks a lot @mlxd

**Review by AmintorDusko — APPROVED** [review]
Looks good. Only left one suggestions.

File: .github/CHANGELOG.md (hunk 2 above)
Comment (mlxd) [reply .github/CHANGELOG.md:24]: That was a hastily written message indeed. Let's push ahead for the GPU fixes, and I'll make it better for release
